File: src/lib/scala/Slate.Common/src/main/scala/slate/common/info/Lang.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
  */

package slate.common.info

/**
 * Represents a host such as a cloud server. e.g. ec2
 * @param name       : name of the language
 * @param home       : home directory of the language
 * @param origin     : origin of the language, e.g. for scala -reference to jre
 * @param versionNum : version of the language
 * @param version    : addition info about architechture for lang ( e.g. 64 bit )
 * @param ext1       : additional information about the language
 */
case class Lang(
                  name        : String = "",
                  home        : String = "",
                  versionNum  : String = "",
                  version     : String = "",
                  origin      : String = "",
                  ext1        : String = ""
               )
{

  def log( callback:(String,String) => Unit) : Unit = {
    callback("name"   , name )
    callback("home"     , home )
    callback("versionNum" , versionNum )
    callback("version"   , version )
    callback("origin", origin )
    callback("ext1"   , ext1 )
  }
}


object Lang
{
  val none = new Lang(
    name     = "none",
    home     = "-",
    versionNum = "",
    version  = "-",
    origin   = "local",
    ext1        = "-"
  )


  def asScala(): Lang =
  {
    new Lang(
      name       = "scala",
      home       = scala.util.Properties.javaHome.replaceAllLiterally("\\", "/"),
      versionNum = scala.util.Properties.versionNumberString,
      version    = scala.util.Properties.javaVersion,
      origin     = "local",
      ext1       = ""
    )
  }
}

File: src/lib/kotlin/slatekit-common/src/main/kotlin/slatekit/common/info/Lang.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.common.info

import slatekit.common.utils.Props

/**
 * Represents a host such as a cloud server. e.g. ec2
 * @param name : name of the language
 * @param home : home directory of the language
 * @param origin : origin of the language, e.g. for scala -reference to jre
 * @param versionNum : version of the language
 * @param version : addition info about architechture for lang ( e.g. 64 bit )
 * @param ext1 : additional information about the language
 */
data class Lang(

    @JvmField
    val name: String,

    @JvmField
    val home: String,

    @JvmField
    val version: String,

    @JvmField
    val vendor: String,

    @JvmField
    val origin: String,

    @JvmField
    val ext1: String
) : Info {

    override fun props():List<Pair<String,String>> = listOf(
            "name"    to name,
            "home"    to home,
            "vendor"  to vendor,
            "version" to version,
            "origin"  to origin,
            "ext1"    to ext1
    )

    companion object {
        @JvmStatic
        val none = Lang(
            name = "none",
            home = "-",
            vendor = "",
            version = "-",
            origin = "local",
            ext1 = "-"
        )

        @JvmStatic
        fun kotlin(): Lang =
            Lang(
                name = "kotlin",
                home = Props.javaHome.replace("\\", "/"),
                version = Props.javaVersion,
                vendor = Props.javaVendor,
                origin = "local",
                ext1 = ""
            )
    }
}
